File: src/thytrader/api/routes/fees.py
"""Read-only Coinbase fee tier and transaction cost presentation."""

from __future__ import annotations

from datetime import UTC, date, datetime
from decimal import Decimal
import logging
from typing import TYPE_CHECKING, Annotated, Literal

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, ConfigDict, Field, field_validator

from thytrader.api.dependencies import get_portfolio_service
from thytrader.exchanges.fee_schedule import suggest_research_fee_rates
from thytrader.portfolio.service import (
    PortfolioService,  # noqa: TC001 - FastAPI resolves dependency at runtime.
)

if TYPE_CHECKING:
    from thytrader.exchanges.fees import FeeProfile

router = APIRouter(prefix="/api/v1/fees", tags=["fees"])
_logger = logging.getLogger(__name__)


class _FrozenResponseModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class FeeProfileResponse(_FrozenResponseModel):
    """Exact fee tier snapshot plus research-only suggested maker/taker rates.

    Dashboard ``maker_fee_rate`` / ``taker_fee_rate`` remain the Coinbase snapshot.
    ``suggested_*`` fields are modeled research defaults mapped through the pinned
    schedule, or null when demo/missing credentials cannot supply a real tier.
    """

    taker_fee_rate: Decimal = Field(ge=Decimal("0"), le=Decimal("1"))
    maker_fee_rate: Decimal = Field(ge=Decimal("0"), le=Decimal("1"))
    usd_volume_30d: Decimal = Field(ge=Decimal("0"))
    fee_tier: str = Field(min_length=1, max_length=64)
    as_of: datetime
    source: Literal["coinbase"]
    suggested_maker_fee_rate: Decimal | None = Field(default=None, ge=Decimal("0"), le=Decimal("1"))
    suggested_taker_fee_rate: Decimal | None = Field(default=None, ge=Decimal("0"), le=Decimal("1"))
    suggestion_source: Literal["coinbase_fee_schedule", "unavailable"]
    suggestion_unavailable_reason: Literal["demo_or_missing_credentials"] | None = None
    suggestion_fee_tier: str | None = Field(default=None, min_length=1, max_length=64)
    suggestion_schedule_tier_id: str | None = Field(default=None, min_length=1, max_length=32)
    suggestion_schedule_version: str | None = Field(default=None, min_length=1, max_length=64)
    suggestion_schedule_as_of: date | None = None
    suggestion_fetched_at: datetime | None = None

    @field_validator("as_of")
    @classmethod
    def require_utc_timezone(cls, value: datetime) -> datetime:
        """Reject naive datetimes to prevent ambiguous fee evaluation times."""
        if value.tzinfo is not UTC:
            raise ValueError("as_of must be timezone-aware UTC")
        return value

    @field_validator("suggestion_fetched_at")
    @classmethod
    def require_utc_suggestion_time(cls, value: datetime | None) -> datetime | None:
        """Reject naive suggestion timestamps; allow null when no suggestion exists."""
        if value is None:
            return None
        if value.tzinfo is not UTC:
            raise ValueError("suggestion_fetched_at must be timezone-aware UTC")
        return value


class FeeErrorDetail(_FrozenResponseModel):
    """Stable redacted error detail for fee service failures."""

    code: Literal["fees_unavailable"]
    message: str


class FeeErrorResponse(_FrozenResponseModel):
    """FastAPI error envelope for fee retrieval failures."""

    detail: FeeErrorDetail


@router.get(
    "",
    response_model=FeeProfileResponse,
    responses={status.HTTP_502_BAD_GATEWAY: {"model": FeeErrorResponse}},
)
async def get_fee_profile(
    portfolio_service: Annotated[PortfolioService, Depends(get_portfolio_service)],
) -> FeeProfileResponse:
    """Return the current 30-day volume and fee tier."""
    try:
        profile = await portfolio_service.get_fee_profile()
    except (RuntimeError, TypeError, ValueError) as error:
        _logger.warning("Fee profile retrieval failed: %s", type(error).__name__)
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail={
                "code": "fees_unavailable",
                "message": "Fee profile is temporarily unavailable.",
            },
        ) from None
    except Exception as error:  # noqa: BLE001
        _logger.warning("Unexpected fee profile error: %s", type(error).__name__)
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail={
                "code": "fees_unavailable",
                "message": "Fee profile is temporarily unavailable.",
            },
        ) from None

    try:
        return _fee_profile_response(profile, demo=portfolio_service.demo)
    except Exception as error:  # noqa: BLE001
        _logger.warning("Failed to serialize fee profile: %s", type(error).__name__)
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail={
                "code": "fees_unavailable",
                "message": "Fee profile is temporarily unavailable.",
            },
        ) from None


def _fee_profile_response(profile: FeeProfile, *, demo: bool) -> FeeProfileResponse:
    """Project a Coinbase snapshot plus research-only suggested maker/taker rates."""
    suggestion = suggest_research_fee_rates(profile=profile, demo=demo)
    return FeeProfileResponse(
        taker_fee_rate=profile.taker_fee_rate,
        maker_fee_rate=profile.maker_fee_rate,
        usd_volume_30d=profile.usd_volume_30d,
        fee_tier=profile.fee_tier,
        as_of=profile.as_of,
        source=profile.source,
        suggested_maker_fee_rate=suggestion.suggested_maker_fee_rate,
        suggested_taker_fee_rate=suggestion.suggested_taker_fee_rate,
        suggestion_source=suggestion.source,
        suggestion_unavailable_reason=suggestion.unavailable_reason,
        suggestion_fee_tier=suggestion.fee_tier,
        suggestion_schedule_tier_id=suggestion.schedule_tier_id,
        suggestion_schedule_version=suggestion.schedule_version,
        suggestion_schedule_as_of=suggestion.schedule_as_of,
        suggestion_fetched_at=suggestion.fetched_at,
    )

File: web/src/lib/fees.ts
/**
 * Typed API client for Coinbase fee tiers and research maker/taker suggestions.
 */

import { compareDecimalStrings } from './portfolio';

export const RESEARCH_FEE_ENGINE_NOTE =
	'These are modeled research assumptions, not observed Coinbase fills. V1 and V2 next-open fills use the taker rate even when the strategy prefers maker.';

export interface FeeProfile {
	taker_fee_rate: string;
	maker_fee_rate: string;
	usd_volume_30d: string;
	fee_tier: string;
	as_of: string;
	source: 'coinbase';
	suggested_maker_fee_rate?: string | null;
	suggested_taker_fee_rate?: string | null;
	suggestion_source?: 'coinbase_fee_schedule' | 'unavailable';
	suggestion_unavailable_reason?: 'demo_or_missing_credentials' | null;
	suggestion_fee_tier?: string | null;
	suggestion_schedule_tier_id?: string | null;
	suggestion_schedule_version?: string | null;
	suggestion_schedule_as_of?: string | null;
	suggestion_fetched_at?: string | null;
}

export interface ResearchFeeSuggestion {
	makerFeeRate: string;
	takerFeeRate: string;
	feeTier: string;
	scheduleTierId: string;
	scheduleVersion: string;
	scheduleAsOf: string;
	fetchedAt: string;
}

export type ResearchFeeFieldSource =
	'loading' | 'unavailable' | 'suggested' | 'stale-suggestion' | 'custom';

export interface FeeErrorResponse {
	detail?: {
		code: string;
		message: string;
	};
}

/**
 * Format `FeeProfile.as_of` for the fees panel, or null when it is absent or invalid.
 */
export function formatFeeProfileAsOf(asOf: string): string | null {
	if (asOf.trim() === '') {
		return null;
	}
	const milliseconds = Date.parse(asOf);
	if (!Number.isFinite(milliseconds)) {
		return null;
	}
	return new Date(milliseconds).toLocaleString();
}

/**
 * Read a schedule-backed research suggestion, or null when the payload is unavailable.
 */
export function readResearchFeeSuggestion(profile: FeeProfile): ResearchFeeSuggestion | null {
	if (profile.suggestion_source !== 'coinbase_fee_schedule') {
		return null;
	}
	const maker = profile.suggested_maker_fee_rate?.trim() ?? '';
	const taker = profile.suggested_taker_fee_rate?.trim() ?? '';
	const feeTier = profile.suggestion_fee_tier?.trim() ?? '';
	const scheduleVersion = profile.suggestion_schedule_version?.trim() ?? '';
	const fetchedAt = profile.suggestion_fetched_at?.trim() ?? '';
	if (
		maker === '' ||
		taker === '' ||
		feeTier === '' ||
		scheduleVersion === '' ||
		fetchedAt === ''
	) {
		return null;
	}
	return {
		makerFeeRate: maker,
		takerFeeRate: taker,
		feeTier,
		scheduleTierId: profile.suggestion_schedule_tier_id?.trim() ?? '',
		scheduleVersion,
		scheduleAsOf: profile.suggestion_schedule_as_of?.trim() ?? '',
		fetchedAt
	};
}

/**
 * Compare two decimal fee-rate strings without binary floating point.
 */
export function feeRatesMatch(left: string, right: string): boolean {
	const trimmedLeft = left.trim();
	const trimmedRight = right.trim();
	if (trimmedLeft === '' || trimmedRight === '') {
		return false;
	}
	try {
		return compareDecimalStrings(trimmedLeft, trimmedRight) === 0;
	} catch {
		return trimmedLeft === trimmedRight;
	}
}

/**
 * Classify research maker/taker fields as suggested, stale, custom, or unavailable.
 */
export function researchFeeFieldSource(input: {
	makerFeeRate: string;
	takerFeeRate: string;
	applied: ResearchFeeSuggestion | null;
	latest: ResearchFeeSuggestion | null;
	loading: boolean;
}): ResearchFeeFieldSource {
	const empty = input.makerFeeRate.trim() === '' && input.takerFeeRate.trim() === '';
	if (empty && input.loading) {
		return 'loading';
	}
	if (empty) {
		return 'unavailable';
	}
	if (
		input.latest !== null &&
		feeRatesMatch(input.makerFeeRate, input.latest.makerFeeRate) &&
		feeRatesMatch(input.takerFeeRate, input.latest.takerFeeRate)
	) {
		return 'suggested';
	}
	if (
		input.applied !== null &&
		feeRatesMatch(input.makerFeeRate, input.applied.makerFeeRate) &&
		feeRatesMatch(input.takerFeeRate, input.applied.takerFeeRate)
	) {
		return 'stale-suggestion';
	}
	return 'custom';
}

/**
 * Honest source chip copy for research maker/taker fields.
 */
export function formatResearchFeeSourceChip(
	source: ResearchFeeFieldSource,
	asOfLabel: string | null
): string {
	if (source === 'loading') {
		return 'Loading fee-tier suggestion…';
	}
	if (source === 'unavailable') {
		return 'Coinbase fee-tier suggestion unavailable. Enter modeled rates.';
	}
	if (source === 'custom') {
		return 'Custom';
	}
	const asOf = asOfLabel !== null ? ` (as of ${asOfLabel})` : '';
	if (source === 'stale-suggestion') {
		return `Suggested from Coinbase fee tier${asOf} — stale`;
	}
	return `Suggested from Coinbase fee tier${asOf}`;
}

/**
 * Prefill empty untouched fields only; never overwrite in-progress edits.
 */
export function shouldPrefillResearchFeeRates(input: {
	makerFeeRate: string;
	takerFeeRate: string;
	touched: boolean;
	suggestion: ResearchFeeSuggestion | null;
}): boolean {
	return (
		input.suggestion !== null &&
		!input.touched &&
		input.makerFeeRate.trim() === '' &&
		input.takerFeeRate.trim() === ''
	);
}

export async function fetchFeeProfile(): Promise<FeeProfile> {
	const response = await fetch('/api/v1/fees', {
		headers: { Accept: 'application/json' }
	});

	if (!response.ok) {
		let message = 'Fee profile is temporarily unavailable.';
		try {
			const errorBody = (await response.json()) as FeeErrorResponse;
			if (errorBody.detail?.message) {
				message = errorBody.detail.message;
			}
		} catch {
			// fallback
		}
		throw new Error(message);
	}

	return (await response.json()) as FeeProfile;
}
